src/v/coproc/reconciliation_backend.cc

```diff
@@ -36,94 +62,112 @@ reconciliation_backend::reconciliation_backend(
+        });
+    } catch (const ss::gate_closed_exception&) {
+        vlog(
+          coproclog.debug,
+          "Gate closed encountered when attempting to start reconciliation "
+          "backend");
+    }
     return ss::now();
 }
 
 ss::future<> reconciliation_backend::stop() {
-    _retry_timer.cancel();
     _as.request_abort();
     _topics.local().unregister_delta_notification(_id_cb);
     return _gate.close();
 }
 
-ss::future<std::vector<reconciliation_backend::update_t>>
-reconciliation_backend::process_events_for_ntp(
+ss::future<> reconciliation_backend::process_updates(
   model::ntp ntp, std::vector<update_t> updates) {
-    std::vector<update_t> retries;
-    for (auto& d : updates) {
-        vlog(coproclog.trace, "executing ntp: {} op: {}", d.ntp, d);
-        auto err = co_await process_update(d);
-        vlog(coproclog.info, "partition operation {} result {}", d, err);
-        if (err == errc::partition_not_exists) {
-            /// In this case the source topic exists but its
-            /// associated partition doesn't, so try again
-            retries.push_back(std::move(d));
+    auto itr = updates.begin();
+    while (itr != updates.end()) {
+        vlog(coproclog.trace, "executing ntp: {} op: {}", ntp, *itr);
+        auto err = co_await process_update(ntp, *itr);
+        vlog(coproclog.info, "partition operation {} result {}", *itr, err);
+        /// This retry mechanism exists to replay events that must be
+        /// executed before progress can be made.
+        ///
+        /// An example is creating a materialized partition. This action
+        /// depends on the cluster::controller_backend having first
+        /// created the source partition. This is an event that MUST have
+        /// occurred first in absolute time otherwise an event to create a
+        /// materialized log with that source as a parent would never have
+        /// been fired.
+        if (err != errc::partition_not_exists) {
+            ++itr;
+        } else {
+            vlog(
+              coproclog.warn,
+              "There were recoverable errors when processing events, "
+              "retrying: {}",
+              *itr);
+            try {
+                co_await ss::sleep_abortable(1s, _as);
+            } catch (const ss::sleep_aborted& e) {
+                vlog(coproclog.debug, "Aborted sleep due to shutdown");
+                break;
+            }
         }
     }
-    co_return retries;
 }
 
-ss::future<> reconciliation_backend::fetch_and_reconcile() {
-    using deltas_cache = decltype(_topic_deltas);
-    auto deltas = std::exchange(_topic_deltas, {});
-    deltas_cache retry_cache;
+ss::future<>
+reconciliation_backend::fetch_and_reconcile(events_cache_t deltas) {
+    if (deltas.empty()) {
+        /// Prevent spin looping when theres nothing to process
+        try {
+            co_await ss::sleep_abortable(1s, _as);
```

Comment:
So that exists (at the begging of `start()`), but since that method isn't futurized a mechanism needs to be in-place to ensure in-order processing of these events.

Previously I was using the trigger as a mechanism to start processing, and if processing is already started, just start a timer to re-start sometime in the future. But I think it might be harder to get that right, as i found some bugs with out of order processing.

With this new implementation, when the delta event arrives, it just places the update in a vector of to-be-processed events, where this main reconciliation fiber will eventually consume it. No timers, no mutexes and a simpler thread of context to understand, i.e. the loop cannot continue until the current update completes without a retryable error.